strawberryfields/tdm/tdmprogram.py

```diff
@@ -0,0 +1,353 @@
+    """Checks the input arguments have consistent dimensions.
+
+    Args:
+        args (Sequence[Sequence]): sequence of sequences specifying the value of the parameters
+        copies (int): number of times the circuit should be run
+
+    """
+    # check if all lists are of equal length
+    param_lengths = [len(param) for param in args]
+    if len(set(param_lengths)) != 1:
+        raise ValueError("Gate-parameter lists must be of equal length.")
+
+    # check copies
+    if not isinstance(copies, int) or copies < 1:
+        raise TypeError("Number of copies must be a positive integer.")
+
+
+def _get_mode_order(num_of_values, N):
+    """Get the order by which the modes were measured"""
+    N = N.copy()
+
+    all_modes = []
+    for i in range(len(N)):
+        ra = list(range(sum(N[:i]), sum(N[: i + 1])))
+        all_modes.append(ra * ceil(max(N) / len(ra)))
+
+    mode_order = [i for j in zip(*all_modes) for i in j]
+    mode_order *= ceil(num_of_values / len(mode_order))
+
+    return mode_order[:num_of_values]
+
+
+def reshape_samples(all_samples, modes, N):
+    """Reshapes the samples dict so that they have the expected correct shape
+
+    Reshapes the samples dict so that they the measured modes are the ones
+    defined to be measured in the circuit instead of being spread over a larger
+    number of modes due to the mode-shifting occuring in TDMProgram.
+
+    Args:
+        all_samples (dict[int, list]): the raw measured samples
+        modes (Sequence[int]): the modes that are measured in the circuit
+        N (Sequence[int]): the number of used modes per belt/spatial modes
+
+    Returns:
+        new_samples (dict[int, list]): the re-shaped samples
+    """
+    num_of_values = len([i for j in all_samples.values() for i in j])
+    mode_order = _get_mode_order(num_of_values, N)
+
+    # go backwards through all_samples and add them into the correct mode
+    new_samples = dict()
+    for i, m in enumerate(mode_order):
+        idx = modes[i % len(N)]
+        if idx not in new_samples:
+            new_samples[idx] = []
+        new_samples[idx].append(all_samples[m].pop(0))
+    return new_samples
+
+
+class TDMProgram(sf.Program):
+    r"""Represents a photonic quantum circuit in the time domain encoding.
+
+    The ``TDMProgam`` class provides a context manager for easily defining
+    a single time-bin of the time domain algorithm. As with the standard
+    :class:`~.Program`, Strawberry Fields operations are appended to the
+    time domain program using the Python-embedded Blackbird syntax.
+
+    Once created, time domain programs can be executed on Strawberry Fields'
+    suite of built-in simulators in an efficient manner, or submitted
+    to be executed on compatible hardware.
+
+    Args:
+        N (int or Sequence[int]): If an integer, the number of concurrent (or 'alive')
+            modes in each time bin. Alternatively, a sequence of integers
+            may be provided, corresponding to
```

Comment:
This sentence ends quite abruptly. I'm guessing that something is missing here... :thinking: